### DISCUSSION TITLE
Logging the unhandledRejection's stacktrace to a database

### DISCUSSION THREAD

**Discussion by AmosSzust** [nodejs/help#5007] [Q&A]
Hi
Despite my searching, I couldn't find a clear explanation about getting the stacktrace of an unhandledRejection. 

When such exception occurred, Node throws a clear and concise stack trace that shows me the exact problematic line (2nd stacktrace line, userService.ts:91):
```
Error: Illegal arguments: string, undefined
at Object.bcrypt.compareSync (/home/appName/public_html/appNameServer/node_modules/bcryptjs/dist/bcrypt.js:265:19)
at /home/appName/public_html/src/services/userServices.ts:91:17
at Generator.next (<anonymous>)
at fulfilled (/home/appName/public_html/appNameServer/services/userServices.js:5:58)
```

I did the following:
```
.on('unhandledRejection', async (reason: any, promise) => {
        console.log(`${new Date().toISOString()}: Reason: ${reason}`);
        await firestoreClient.handleError(new AppError(true, `Reason: ${reason}`, 500));
        process.exit(1);
    })
```
AppError extends Error. The stack trace in the database contains the following which leads me to the `unhandledRejection` line and not to the original problematic line like in the above stacktrace: `at /home/appName/public_html/src/appNameServer.ts:66:43`
How can I have the nice and clear stacktrace with the relevant info so I can log it to my database?
Thanks

**Comment by bnoordhuis** [comment]
You'll want to log `reason`, not `new AppError(...)`.

Keep in mind `reason` is _usually_ an instance of `Error` but not always.

**Reply by AmosSzust** [reply]
When I logged `reason`, I only got the error string itself, not the whole object/stack.

Eventually I logged `reason.stack` which includes what I needed but it is pointing to the js file/line and not the ts file/line (like node is showing in its own logging when failing). Not perfect but enough for my needs.

If there is a way to get the stack of the ts file, it would have been better.

**Reply by bnoordhuis** [reply]
Maybe`--enable-source-maps` helps? https://nodejs.org/docs/latest-v18.x/api/cli.html#--enable-source-maps

**Reply by AmosSzust** [reply]
I'm using pm2 to run and it says here: https://pm2.keymetrics.io/docs/usage/source-map-support/ that it recognizes filename.js.map (I have this file for each js file) automatically and it throws the stack accordingly. This is exactly what I see in `pm2 logs` but internally, in `unhandledRejection` catcher, this is not the case.
